File: kernel/nodejs/src/bundle/module-manifest.ts
import { parseLayerIndex, type ArtifactLayer } from "@telorun/analyzer";
import { defaultCustomTags } from "@telorun/templating";
import { parseAllDocuments, type Document } from "yaml";

const OWNER_KINDS = new Set(["Telo.Application", "Telo.Library"]);

/** Parse a manifest's YAML documents, tolerating its `!cel` / `!ref` tags. */
export function parseManifestDocs(text: string): Document[] {
  return parseAllDocuments(text, { customTags: defaultCustomTags() });
}

/** The owner document of a manifest — the single `Telo.Application` /
 *  `Telo.Library` doc that carries its identity, `files:` / `assets:`, and the
 *  published `layers:` index. The one place that selects it, shared by every
 *  reader/writer. */
export function findOwnerDoc(docs: Document[]): Document | undefined {
  return docs.find((d) => {
    const kind = (d.toJSON() as { kind?: string } | null)?.kind;
    return kind !== undefined && OWNER_KINDS.has(kind);
  });
}

/** The owner-doc fields the transports read: identity for the publish location,
 *  and payload markers for artifact fetch. */
export interface OwnerManifest {
  name?: string;
  version?: string;
  /** The published layer index, absent on an unpublished manifest. Parsed and
   *  validated through the analyzer's shared model, so a malformed index is a
   *  hard `LayerIndexError` rather than a silently ignored field. */
  layers?: ArtifactLayer[];
  /** Ordered `.gitignore`-style patterns the author claimed as the lazily
   *  materialized `assets` layer. */
  assetPatterns: string[];
  /** True when the owner doc declares a non-empty `files:` list. */
  declaresFiles: boolean;
  /** Descriptive provenance a transport projects into its backend's metadata
   *  (OCI annotations). Never used to address the artifact. */
  description?: string;
  repository?: string;
  license?: string;
  documentation?: string;
}

/** Read the owner doc's identity + payload fields from a manifest, parsing once
 *  (never regex-scraping). The single source both transports call for the
 *  module's name and version, its layer index, and payload detection. */
export function readOwnerManifest(text: string): OwnerManifest {
  const owner = findOwnerDoc(parseManifestDocs(text));
  const parsed = owner?.toJSON() as
    | {
        metadata?: Record<string, unknown>;
        files?: unknown;
        assets?: unknown;
        layers?: unknown;
      }
    | undefined;
  const md = parsed?.metadata ?? {};
  const str = (v: unknown): string | undefined => (typeof v === "string" ? v : undefined);
  const patterns = (v: unknown): string[] =>
    Array.isArray(v) ? v.filter((p): p is string => typeof p === "string") : [];
  return {
    name: str(md.name),
    version: str(md.version),
    layers: parsed?.layers === undefined ? undefined : parseLayerIndex(parsed.layers),
    assetPatterns: patterns(parsed?.assets),
    declaresFiles: Array.isArray(parsed?.files) && parsed.files.length > 0,
    description: str(md.description),
    repository: str(md.repository),
    license: str(md.license),
    documentation: str(md.documentation),
  };
}

File: kernel/rust/src/bundle/module_manifest.rs
//! The owner document of a manifest and the fields the artifact path reads off
//! it. Mirrors `../../../nodejs/src/bundle/module-manifest.ts`.
//!
//! Reads the documents the analyzer's loader already parsed rather than
//! re-parsing text, since every caller holds a `LoadedFile`. Of the Node
//! `OwnerManifest` the layer index and the `sources:` block have consumers here:
//! asset patterns, `files:`, provenance, library candidates and `native:` serve
//! publishing, sibling-library resolution and native-file lookup, none of which
//! this kernel implements.

use serde_json::Value;
use telo_analyzer::artifact_layer_index::{parse_layer_index, ArtifactLayer, LayerIndexError};
use telo_analyzer::kind_of;
use telo_analyzer::source_entries::{read_module_sources, ModuleSources};
use telo_analyzer::system_kinds::is_module_document;

/// The owner-doc fields the kernel reads.
pub struct OwnerManifest {
    /// The published layer index, absent on an unpublished manifest. A malformed
    /// index is a hard error, never a silently ignored field.
    pub layers: Option<Vec<ArtifactLayer>>,
    /// Where each staged file of a source checkout comes from, and its pin.
    pub sources: ModuleSources,
}

/// The single `Telo.Application` / `Telo.Library` document carrying the
/// module's identity and its published `layers:` index.
fn find_owner_doc<'a>(documents: impl IntoIterator<Item = &'a Value>) -> Option<&'a Value> {
    documents
        .into_iter()
        .find(|document| kind_of(document).is_some_and(is_module_document))
}

pub fn read_owner_manifest<'a>(
    documents: impl IntoIterator<Item = &'a Value>,
) -> Result<OwnerManifest, LayerIndexError> {
    let owner = find_owner_doc(documents);
    let layers = owner
        .and_then(|owner| owner.get("layers"))
        .map(|layers| parse_layer_index(layers, "layers"))
        .transpose()?;
    let sources = owner.map(read_module_sources).unwrap_or_default();
    Ok(OwnerManifest { layers, sources })
}
